### PR TITLE
Email notifications toggle API endpoint

### PR DESCRIPTION
**Pull request by thedae**
This PR enables two new API endpoints to enable/disable email notifications (such as DO notifications):

**GET** `/api/v3/email_notifications?api_key=abcd`
Returns a JSON with the list of email notifications and the current status, for example:
```
{
    "notifications": {
        "do_subscriptions": true
    }
}
```

**PUT** `/api/v3/email_notifications?api_key=abcd`
Create or update a given list of notifications. This endpoint needs a JSON payload as request parameter, for example:
```
{
    "notifications": {
        "do_subscriptions": false,
        "some_other_subscription": true
    }
}
```
You don't need to always provide the complete list of notifications to update them.

### CODE REVIEW

File: app/models/carto/user_email_notification.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        user_id: user_id,
+        notification: NOTIFICATION_DO_SUBSCRIPTIONS,
+        enabled: true
+      )
+    end
+
+    private
+
+    def valid_notification
+      errors.add(:notification, 'Invalid notification') unless VALID_NOTIFICATIONS.include?(notification)
```
Comment (rafatower) [inline app/models/carto/user_email_notification.rb:23]: nitpicking: you can use the `inclusion` validation helper: https://guides.rubyonrails.org/v4.2/active_record_validations.html#inclusion

File: app/controllers/carto/api/email_notifications_controller.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+module Carto
+  module Api
+    class EmailNotificationsController < ::Api::ApplicationController
+
+      ssl_required :update
+      before_action :load_notifications, only: [:show, :update]
+
+      rescue_from StandardError, with: :rescue_from_standard_error
+      rescue_from Carto::LoadError, with: :rescue_from_carto_error
```
Comment (rafatower) [inline app/controllers/carto/api/email_notifications_controller.rb:11]: from my ignorance, why is this needed?

File: app/controllers/carto/api/email_notifications_controller.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+require_dependency 'carto/uuidhelper'
```
Comment (rafatower) [inline app/controllers/carto/api/email_notifications_controller.rb:1]: I think this is not needed

File: db/migrate/20201111102307_create_user_email_notifications.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
+require 'carto/db/migration_helper'
+
+include Carto::Db::MigrationHelper
+
+migration(
+  proc do
+    create_table :user_email_notifications do
+      Uuid :id, primary_key: true, default: Sequel.lit('uuid_generate_v4()')
+      foreign_key :user_id, :users, type: :uuid, null: false, on_delete: :cascade
```
Comment (rafatower) [inline db/migrate/20201111102307_create_user_email_notifications.rb:9]: (I was wondering if this `on_delete: :cascade` collides with user model `dependent: :destroy`, but I see there's even a test for it :+1: )

**Review by rafatower — APPROVED** [review]
LGTM, just left minor comments, nothing blocking :muscle:

File: app/controllers/carto/api/email_notifications_controller.rb (hunk 2 above)
Comment (thedae) [reply app/controllers/carto/api/email_notifications_controller.rb:11]: Probably a leftover of a copy&paste from other controller, let me check and delete it

File: app/models/carto/user_email_notification.rb (hunk 1 above)
Comment (thedae) [reply app/models/carto/user_email_notification.rb:23]: I'm having some issues applying this suggestion: in the spec file now I have to mock the constant instead of the `valid_notification` method, and I can't figure out a good solution to achieve that

Comment (rafatower) [reply app/models/carto/user_email_notification.rb:23]: then I guess you have two options: rely on rails doing the validation and not testing it, or keep it "as is" :)

(no big deal)

Comment (thedae) [reply app/models/carto/user_email_notification.rb:23]: I prefer to test it, even if the validation method is not the ideal, so I'm going to keep it as is

File: app/controllers/carto/api/email_notifications_controller.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
+module Carto
+  module Api
+    class EmailNotificationsController < ::Api::ApplicationController
+
+      ssl_required :update
+      before_action :load_notifications, only: [:show, :update]
```
Comment (amiedes) [inline app/controllers/carto/api/email_notifications_controller.rb:6]: I think the `, only: [:show, :update]` is not needed, as for the time being this controller only has those two actions.

File: app/models/carto/user.rb [hunk 6]
```diff
@@ -293,6 +294,28 @@ def dbdirect_effective_ip
     reload.dbdirect_ip
   end
 
+  # @param [Hash] notifications_hash A notification list with this format: {"notif_a" => true, "notif_b" => false}
+  def email_notifications=(notifications_hash)
```
Comment (amiedes) [inline app/models/carto/user.rb:298]: Keep in mind you're overriding the behavior of the `email_notifications=` which Rails creates automatically when you declare the association:

```
has_many :email_notifications
```

Not sure if this might cause issues. The `build` method might come in handy if you want to try it: https://apidock.com/rails/ActiveRecord/Associations/CollectionProxy/build

**Review by amiedes — APPROVED** [review]
I'm leaving a couple of comments 👀

File: app/controllers/carto/api/email_notifications_controller.rb (hunk 5 above)
Comment (thedae) [reply app/controllers/carto/api/email_notifications_controller.rb:6]: You're right, it's useless
